dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/JdbcEnrollmentAnalyticsManager.java

```diff
@@ -235,25 +316,32 @@ else if ( params.isOrganisationUnitMode( OrganisationUnitSelectionMode.CHILDREN
         return sql;
     }
 
-    protected String getBoundedDataValueSelectSql( String programStageUid, String dataElementUid, Date reportingStartDate,
-        Date reportingEndDate, ProgramIndicator programIndicator )
+        /**
+     * Returns an encoded column name wrapped in lower directive if not numeric
+     * or boolean.
+     *
+     * @param item the {@link QueryItem}.
+     */
+    @Override
+    protected String getColumn( QueryItem item )
     {
-        if ( programIndicator.hasNonDefaultBoundaries() && programIndicator.hasEventBoundary() )
+        String colName = item.getItemName();
+
+        if ( item.hasProgramStage() )
         {
-            String eventTableName = "analytics_event_" + programIndicator.getProgram().getUid();
-            String columnName = "\"" + dataElementUid + "\"";
-            return "(select " + columnName + " from " + eventTableName + " where " + eventTableName +
-                ".pi = enrollmenttable.pi and " + columnName + " is not null " +
-                ( programIndicator.getEndEventBoundary() != null ? ( "and " +
-                statementBuilder.getBoundaryCondition( programIndicator.getEndEventBoundary(), programIndicator, reportingStartDate, reportingEndDate ) +
-                    " ") : "" ) + (programIndicator.getStartEventBoundary() != null ? ("and " +
-                statementBuilder.getBoundaryCondition( programIndicator.getStartEventBoundary(), programIndicator, reportingStartDate, reportingEndDate ) +
-                    " ") : "" ) + "and ps = '" + programStageUid + "' " + "order by executiondate " + "desc limit 1 )";
+            colName = quote( colName );
+            Assert.isTrue( item.hasProgram(), "Can not query item with program stage but no program:" + item.getItemName() );
+            String eventTableName = "analytics_event_" + item.getProgram().getUid();
+            return "(select " + ( item.isText() ? "lower(" + colName + ")" : colName ) + " from " +
```

Comment:
Is this value safe for sql injection?

Reply:
For the queryitem to be created, its name has to be identified as either a data element uid, program indicator uid, attribute uid and so on. We try to fetch the data element/program indicator/attribute by ID, and only if it is found to be a valid ID we create a QueryItem object.

Here is my hacking attempts:
{
    "httpStatus": "Conflict",
    "httpStatusCode": 409,
    "status": "ERROR",
    "message": "Dimension: edqlb\"; drop database; \"ukwRfQ.vANAXwtLwcT did not have a valid program stage"
}

{
    "httpStatus": "Conflict",
    "httpStatusCode": 409,
    "status": "ERROR",
    "message": "Item identifier does not reference any data element, attribute or indicator part of the program: vANAXw\"; drop database; \"tLwcT"
}